File: clients/java/src/test/java/io/zeebe/client/workflow/SetVariablesTest.java
/*
 * Copyright © 2017 camunda services GmbH (info@camunda.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.zeebe.client.workflow;

import static io.zeebe.client.util.JsonUtil.fromJsonAsMap;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.entry;

import io.zeebe.client.api.command.ClientException;
import io.zeebe.client.api.response.SetVariablesResponse;
import io.zeebe.client.util.ClientTest;
import io.zeebe.client.util.StringUtil;
import io.zeebe.gateway.protocol.GatewayOuterClass.SetVariablesRequest;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.time.Duration;
import java.util.Collections;
import java.util.Map;
import org.junit.Test;

public final class SetVariablesTest extends ClientTest {

  @Test
  public void shouldCommandWithVariablesAsString() {
    // given
    gatewayService.onSetVariablesRequest(345);
    final String variables = "{\"key\": \"val\"}";

    // when
    final SetVariablesResponse response =
        client.newSetVariablesCommand(123).variables(variables).send().join();

    // then
    assertThat(response).isNotNull();
    assertThat(response.getKey()).isEqualTo(345);

    final SetVariablesRequest request = gatewayService.getLastRequest();
    assertThat(request.getElementInstanceKey()).isEqualTo(123);
    assertThat(fromJsonAsMap(request.getVariables())).containsOnly(entry("key", "val"));

    rule.verifyDefaultRequestTimeout();
  }

  @Test
  public void shouldCommandWithVariablesAsStream() {
    // given
    final String variables = "{\"key\": \"val\"}";
    final InputStream variablesStream = new ByteArrayInputStream(StringUtil.getBytes(variables));

    // when
    client.newSetVariablesCommand(123).variables(variablesStream).send().join();

    // then
    final SetVariablesRequest request = gatewayService.getLastRequest();
    assertThat(fromJsonAsMap(request.getVariables())).containsOnly(entry("key", "val"));
  }

  @Test
  public void shouldCommandWithVariablesAsMap() {
    // given
    final Map<String, Object> variablesMap = Collections.singletonMap("key", "val");

    // when
    client.newSetVariablesCommand(123).variables(variablesMap).send().join();

    // then
    final SetVariablesRequest request = gatewayService.getLastRequest();
    assertThat(fromJsonAsMap(request.getVariables())).containsOnly(entry("key", "val"));
  }

  @Test
  public void shouldCommandWithVariablesAsObject() {
    // given
    final Document document = new Document("val");

    // when
    client.newSetVariablesCommand(123).variables(document).send().join();

    // then
    final SetVariablesRequest request = gatewayService.getLastRequest();
    assertThat(fromJsonAsMap(request.getVariables())).containsOnly(entry("key", "val"));
  }

  @Test
  public void shouldRaiseExceptionOnError() {
    // given
    gatewayService.errorOnRequest(
        SetVariablesRequest.class, () -> new ClientException("Invalid request"));

    // when
    assertThatThrownBy(() -> client.newSetVariablesCommand(123).variables("[]").send().join())
        .isInstanceOf(ClientException.class)
        .hasMessageContaining("Invalid request");
  }

  @Test
  public void shouldSetRequestTimeout() {
    // given
    final Duration requestTimeout = Duration.ofHours(124);

    // when
    client.newSetVariablesCommand(123).variables("{}").requestTimeout(requestTimeout).send().join();

    // then
    rule.verifyRequestTimeout(requestTimeout);
  }

  static class Document {
    public final String key;

    Document(final String key) {
      this.key = key;
    }
  }
}

File: clients/go/pkg/commands/set_variables_test.go
// Copyright © 2018 Camunda Services GmbH (info@camunda.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package commands

import (
	"context"
	"github.com/camunda/zeebe/clients/go/internal/mock_pb"
	"github.com/camunda/zeebe/clients/go/internal/utils"
	"github.com/camunda/zeebe/clients/go/pkg/pb"
	"github.com/golang/mock/gomock"
	"testing"
)

func TestSetVariablesCommandWithVariablesFromString(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	client := mock_pb.NewMockGatewayClient(ctrl)

	variables := "{\"foo\":\"bar\"}"

	request := &pb.SetVariablesRequest{
		ElementInstanceKey: 123,
		Variables:          variables,
	}
	stub := &pb.SetVariablesResponse{
		Key: 523,
	}

	client.EXPECT().SetVariables(gomock.Any(), &utils.RPCTestMsg{Msg: request}).Return(stub, nil)

	command := NewSetVariablesCommand(client, func(context.Context, error) bool { return false })

	variablesCommand, err := command.ElementInstanceKey(123).VariablesFromString(variables)
	if err != nil {
		t.Error("Failed to set variables: ", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), utils.DefaultTestTimeout)
	defer cancel()

	response, err := variablesCommand.Send(ctx)

	if err != nil {
		t.Errorf("Failed to send request")
	}

	if response != stub {
		t.Errorf("Failed to receive response")
	}
}

func TestSetVariablesCommandWithVariablesFromStringer(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	client := mock_pb.NewMockGatewayClient(ctrl)

	variables := "{\"foo\":\"bar\"}"

	request := &pb.SetVariablesRequest{
		ElementInstanceKey: 123,
		Variables:          variables,
	}
	stub := &pb.SetVariablesResponse{
		Key: 523,
	}

	client.EXPECT().SetVariables(gomock.Any(), &utils.RPCTestMsg{Msg: request}).Return(stub, nil)

	command := NewSetVariablesCommand(client, func(context.Context, error) bool { return false })

	variablesCommand, err := command.ElementInstanceKey(123).VariablesFromStringer(DataType{Foo: "bar"})
	if err != nil {
		t.Error("Failed to set variables: ", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), utils.DefaultTestTimeout)
	defer cancel()

	response, err := variablesCommand.Send(ctx)

	if err != nil {
		t.Errorf("Failed to send request")
	}

	if response != stub {
		t.Errorf("Failed to receive response")
	}
}

func TestSetVariablesCommandWithVariablesFromObject(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	client := mock_pb.NewMockGatewayClient(ctrl)

	variables := "{\"foo\":\"bar\"}"

	request := &pb.SetVariablesRequest{
		ElementInstanceKey: 123,
		Variables:          variables,
	}
	stub := &pb.SetVariablesResponse{
		Key: 523,
	}

	client.EXPECT().SetVariables(gomock.Any(), &utils.RPCTestMsg{Msg: request}).Return(stub, nil)

	command := NewSetVariablesCommand(client, func(context.Context, error) bool { return false })

	variablesCommand, err := command.ElementInstanceKey(123).VariablesFromObject(DataType{Foo: "bar"})
	if err != nil {
		t.Error("Failed to set variables: ", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), utils.DefaultTestTimeout)
	defer cancel()

	response, err := variablesCommand.Send(ctx)

	if err != nil {
		t.Errorf("Failed to send request")
	}

	if response != stub {
		t.Errorf("Failed to receive response")
	}
}

func TestSetVariablesCommandWithVariablesFromObjectOmitempty(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	client := mock_pb.NewMockGatewayClient(ctrl)

	variables := "{}"

	request := &pb.SetVariablesRequest{
		ElementInstanceKey: 123,
		Variables:          variables,
	}
	stub := &pb.SetVariablesResponse{
		Key: 523,
	}

	client.EXPECT().SetVariables(gomock.Any(), &utils.RPCTestMsg{Msg: request}).Return(stub, nil)

	command := NewSetVariablesCommand(client, func(context.Context, error) bool { return false })

	variablesCommand, err := command.ElementInstanceKey(123).VariablesFromObject(DataType{Foo: ""})
	if err != nil {
		t.Error("Failed to set variables: ", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), utils.DefaultTestTimeout)
	defer cancel()

	response, err := variablesCommand.Send(ctx)

	if err != nil {
		t.Errorf("Failed to send request")
	}

	if response != stub {
		t.Errorf("Failed to receive response")
	}
}

func TestSetVariablesCommandWithVariablesFromObjectIgnoreOmitempty(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	client := mock_pb.NewMockGatewayClient(ctrl)

	variables := "{\"foo\":\"\"}"

	request := &pb.SetVariablesRequest{
		ElementInstanceKey: 123,
		Variables:          variables,
	}
	stub := &pb.SetVariablesResponse{
		Key: 523,
	}

	client.EXPECT().SetVariables(gomock.Any(), &utils.RPCTestMsg{Msg: request}).Return(stub, nil)

	command := NewSetVariablesCommand(client, func(context.Context, error) bool { return false })

	variablesCommand, err := command.ElementInstanceKey(123).VariablesFromObjectIgnoreOmitempty(DataType{Foo: ""})
	if err != nil {
		t.Error("Failed to set variables: ", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), utils.DefaultTestTimeout)
	defer cancel()

	response, err := variablesCommand.Send(ctx)

	if err != nil {
		t.Errorf("Failed to send request")
	}

	if response != stub {
		t.Errorf("Failed to receive response")
	}
}

func TestSetVariablesCommandWithVariablesFromMap(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	client := mock_pb.NewMockGatewayClient(ctrl)

	variables := "{\"foo\":\"bar\"}"
	variablesMap := make(map[string]interface{})
	variablesMap["foo"] = "bar"

	request := &pb.SetVariablesRequest{
		ElementInstanceKey: 123,
		Variables:          variables,
	}
	stub := &pb.SetVariablesResponse{
		Key: 523,
	}

	client.EXPECT().SetVariables(gomock.Any(), &utils.RPCTestMsg{Msg: request}).Return(stub, nil)

	command := NewSetVariablesCommand(client, func(context.Context, error) bool { return false })

	variablesCommand, err := command.ElementInstanceKey(123).VariablesFromMap(variablesMap)
	if err != nil {
		t.Error("Failed to set variables: ", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), utils.DefaultTestTimeout)
	defer cancel()

	response, err := variablesCommand.Send(ctx)

	if err != nil {
		t.Errorf("Failed to send request")
	}

	if response != stub {
		t.Errorf("Failed to receive response")
	}
}
